tests/atparse.bash

```diff
@@ -1,42 +1,106 @@
-    local __set_x
-    [ -o xtrace ] && __set_x='set -x' || __set_x='set +x'
-    set +x
+    # Usage:
+    #     source atparse.bash # defines the "atparse" function; only do this once
+    #     atparse [ var1=value1 [ var2=value2 [...] ] ] < input_file > output_file
+    # This function filters text from stdin to stdout. It scans for text sequences like:
+    #    @[varname]
+    # And replaces them with the value of the corresponding ${varname} variable.
+    # You can provide variables that are not set in bash by providing them on the command line.
+    # If set -u is enabled, it will exit the process when a variable is empty or undefined via set -u.
+
     # Use __ in names to avoid clashing with variables in {var} blocks.
-    local __text __before __after __during
+    local __text # current line of text being parsed, or the current command-line argument being parsed
+    local __before # all text before the next @[...] option
+    local __after # all text after the next @[...] option
+    local __during # the contents of the @[...] option, including the @[ and ]
+    local __set_x=":" # will be "set -x" if the calling script had that option enabled
+    local __set_u=":" # will be "set -u" if the calling script had that option enabled
+    local __set_e=":" # will be "set -e" if the calling script had that option enabled
+    local __abort_on_undefined=NO # YES = script should abort if a variable is undefined, NO otherwise
+
+    # Ensure "set -x -e -u" are all inactive, but remember if they
+    # were active so we can reset them later.
+    if [[ -o xtrace ]] ; then
+	__set_x="set -x"
+    fi
+    if [[ -o errexit ]] ; then
+	__set_e="set -e"
+    fi
+    if [[ -o nounset ]] ; then
+	__set_u="set -u"
+	__abort_on_undefined=YES
+    fi
+    set +eux
+
+    # Allow setting variables on the atparse command line rather than the environment.
+    # They will be local variables in this function.
     for __text in "$@" ; do
         if [[ $__text =~ ^([a-zA-Z][a-zA-Z0-9_]*)=(.*)$ ]] ; then
             eval "local ${BASH_REMATCH[1]}"
             eval "${BASH_REMATCH[1]}="'"${BASH_REMATCH[2]}"'
         else
-            echo "ERROR: Ignoring invalid argument $__text\n" 1>&2
+            echo "ERROR: Ignoring invalid argument $__text" 1>&2
         fi
     done
-    while IFS= read -r __text ; do
+
+    # Loop over all lines of text.
+    while [[ 1 == 1 ]] ; do
+        # Read the next line of text. This will "fail" if no more text
+        # is left OR if the last line lacks an end-of-line character.
+        read -d '' -r __text
+
+        # Stop when "read" reports it is done ($? -ne 0) AND the text is
+        # non-empty (! -n "$__text"). This ensures we read the final line
+        # even if it lacks an end-of-line character.
+        if [[ $? -ne 0 ]] ; then
+            if [[ -n "$__text" ]] ; then
+                # Text remained, but it had no end-of-line.
+		:
+            else
+                break
+            fi
+        fi
```

Comment:
This section of code looks like this in my editor:

![image](https://github.com/ufs-community/ufs-weather-model/assets/48258889/36809224-f597-4a1a-8348-92a33e360509)

See how ':' is not properly indented. This is because it's indented with tabs. Your editor interprets tab as 8 spaces, mine as 4. Please replace tabs with spaces.

Reply:
It seems my emacs settings are not properly carrying over to Hera Rocky.

Reply:
And GNU's new Savannah server is blocked by the firewall, so I can't update /contrib/emacs with a version that works on Rocky.

Reply:
I replaced tabs with spaces in rt.sh and atparse.bash.

Reply:
Cool!